source/binding/Bitstream.cpp

```diff
@@ -0,0 +1,656 @@
+
+    return result;
+}
+
+/// Performs unpack operation of streaming concatenation target on a bit-stream.
+static bool unpackConcatenation(const StreamingConcatenationExpression& lhs, PackIterator& iter,
+                                const PackIterator iterEnd, bitwidth_t& bitOffset,
+                                size_t& dynamicSize, EvalContext& context,
+                                SmallVector<ConstantValue>* dryRun = nullptr) {
+    for (auto stream : lhs.streams()) {
+        if (stream->kind == ExpressionKind::Streaming) {
+            auto concat = stream->as<StreamingConcatenationExpression>();
+            if (dryRun || !concat.sliceSize) {
+                if (!unpackConcatenation(stream->as<StreamingConcatenationExpression>(), iter,
+                                         iterEnd, bitOffset, dynamicSize, context, dryRun))
+                    return false;
+            }
+
+            // A dry run collects rvalue without storing lvalue
+            auto dynamicSizeSave = dynamicSize;
+            SmallVectorSized<ConstantValue, 8> toBeOrdered;
+            if (!unpackConcatenation(stream->as<StreamingConcatenationExpression>(), iter, iterEnd,
+                                     bitOffset, dynamicSize, context, &toBeOrdered))
+                return false;
+
+            // Re-order to a new rvalue with the slice size
+            ConstantValue cv = std::vector(toBeOrdered.begin(), toBeOrdered.end());
+            auto rvalue = Bitstream::reOrder(std::move(cv), concat.sliceSize, cv.bitstreamWidth());
+            SmallVectorSized<ConstantValue*, 8> packed;
+            packBitstream(rvalue, packed);
+
+            // A real pass stores lvalue from new rvalue
+            auto iterConcat = packed.cbegin();
+            bitwidth_t bit = 0;
+            if (!unpackConcatenation(concat, iterConcat, packed.cend(), bit, dynamicSizeSave,
+                                     context))
+                return false;
+            ASSERT(dynamicSizeSave == dynamicSize);
+            ASSERT(iterConcat == packed.cend() && !bit);
+        }
+        else {
+            auto rvalue = unpackBitstream(*stream->type, iter, iterEnd, bitOffset, dynamicSize);
+            if (dryRun)
+                dryRun->emplace(std::move(rvalue));
+            else {
+                LValue lvalue = stream->evalLValue(context);
+                if (!lvalue)
+                    return false;
+                lvalue.store(rvalue);
+            }
+        }
+    }
+    return true;
+}
+
+ConstantValue Bitstream::evaluateTarget(const StreamingConcatenationExpression& lhs,
+                                        const Expression& rhs, EvalContext& context) {
+    ConstantValue rvalue = rhs.eval(context);
+    if (!rvalue)
+        return nullptr;
+
+    auto srcSize = rvalue.bitstreamWidth();
+    auto targetWidth = lhs.bitstreamWidth();
+    size_t dynamicSize = 0;
+    if (rhs.kind == ExpressionKind::Streaming) { // srcSize == targetWidth + dynamicSize
```

Comment:
Comments like this are confusing to me... is it asserting that the expression is true? Or is it saying that you're about to check whether it is true or not, and then issue an error if not?

Reply:
The intension is to check the sizes. I improve the comments to
```c++
    if (rhs.kind == ExpressionKind::Streaming) {
        // Check srcSize == targetWidth + dynamicSize, then issue an error if not
    ...
    else {
        // Check srcSize >= targetWidth + dynamicSize, then issue an error if not
```